### DISCUSSION TITLE
Get state inside build method

### DISCUSSION THREAD

**Discussion by MkamaMakoko** [rrousselGit/riverpod#1958] [Q&A] [answered]
Is there a safe way to obtain a state inside build method?
Or if the initial state has not been initialized, is there a way to know it?
Thanks.

**Comment by rrousselGit** [comment, marked as answer]
Do you have a more specific question? I don't get what this is about

**Reply by MkamaMakoko** [reply]
I'm trying to update my state by copying the previous state inside the build method, the problem is, I get an error if I do that. For example

```
StateA build(){
final stateB=ref.watch(providerB)
state=state.copyWith(value1: stateB.value1);
return StataeA(value1: 0, value2: 0);
}
```

My provider is like that, I need to update the state inside the build method because that is where I can safely call ref.watch method. But because there is no initial state in the first run of the build method, I get an error when I try to copy the previous state.

**Reply by rrousselGit** [reply]
This code doesn't make sense, as your state would be reset right after the `state=` to `StataeA(value1: 0, value2: 0)`, rendering the `state=` useless.

But feel free to do something like:

```dart
bool didBuild = false;

@override
build() {
  if (didBuild) {
    state = state.copyWith(...)  
  }
  didBuild = true;
}
```

**Reply by MkamaMakoko** [reply]
> This code doesn't make sense, as your state would be reset right after the `state=` to `StataeA(value1: 0, value2: 0)`, rendering the `state=` useless.
> 
> But feel free to do something like:
> 
> ```dart
> bool didBuild = false;
> 
> @override
> build() {
>   if (didBuild) {
>     state = state.copyWith(...)  
>   }
>   didBuild = true;
> }
> ```

Thank you, I've got the idea
